crates/turbopack-ecmascript/src/references/external_module.rs

```diff
@@ -0,0 +1,243 @@
+    },
+    references::async_module::{AsyncModule, OptionAsyncModule},
+    utils::StringifyJs,
+    EcmascriptModuleContent, EcmascriptOptions,
+};
+
+#[turbo_tasks::function]
+fn layer() -> Vc<RcStr> {
+    Vc::cell("external".into())
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, TraceRawVcs, TaskInput)]
+pub enum CachedExternalType {
+    CommonJs,
+    EcmaScriptViaRequire,
+    EcmaScriptViaImport,
+}
+
+impl Display for CachedExternalType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CachedExternalType::CommonJs => write!(f, "cjs"),
+            CachedExternalType::EcmaScriptViaRequire => write!(f, "esm_require"),
+            CachedExternalType::EcmaScriptViaImport => write!(f, "esm_import"),
+        }
+    }
+}
+
+#[turbo_tasks::value]
+pub struct CachedExternalModule {
+    pub request: RcStr,
+    pub external_type: CachedExternalType,
+}
+
+#[turbo_tasks::value_impl]
+impl CachedExternalModule {
+    #[turbo_tasks::function]
+    pub fn new(request: RcStr, external_type: CachedExternalType) -> Vc<Self> {
+        Self::cell(CachedExternalModule {
+            request,
+            external_type,
+        })
+    }
+
+    #[turbo_tasks::function]
+    pub fn content(&self) -> Result<Vc<EcmascriptModuleContent>> {
+        let mut code = RopeBuilder::default();
+
+        if self.external_type == CachedExternalType::EcmaScriptViaImport {
+            writeln!(
+                code,
+                "const mod = await __turbopack_external_import__({});",
+                StringifyJs(&self.request)
+            )?;
+        } else {
+            writeln!(
+                code,
+                "const mod = __turbopack_external_require__({});",
+                StringifyJs(&self.request)
+            )?;
+        }
+
+        writeln!(code)?;
+
+        if self.external_type == CachedExternalType::CommonJs {
+            writeln!(code, "module.exports = mod;")?;
+        } else {
+            writeln!(code, "__turbopack_export_namespace__(mod);")?;
+        }
+
+        Ok(EcmascriptModuleContent {
+            inner_code: code.build(),
+            source_map: None,
+            is_esm: self.external_type != CachedExternalType::CommonJs,
+        }
+        .cell())
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl Module for CachedExternalModule {
+    #[turbo_tasks::function]
+    fn ident(&self) -> Vc<AssetIdent> {
+        let fs = VirtualFileSystem::new_with_name("externals".into());
+
+        AssetIdent::from_path(fs.root())
+            .with_layer(layer())
+            .with_modifier(Vc::cell(self.request.clone()))
+            .with_modifier(Vc::cell(self.external_type.to_string().into()))
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl Asset for CachedExternalModule {
+    #[turbo_tasks::function]
+    fn content(&self) -> Vc<AssetContent> {
+        AssetContent::file(FileContent::NotFound.cell())
```

Comment:
Is this correct? If it's expected to be not called, I think it's better to use `unreachable!()`

Reply:
done